### PR TITLE
Don't erase env variable when using cpu device

### LINKED ISSUE #2323: Importing Equilibrium removes visible CUDA devices
Hello, I am trying to use DESC as a CPU equilibrium solver with GPU particle tracing code. When I import the Equilibrium object, `CUDA_VISIBLE_DEVICES` gets set to `""`, which my CUDA kernel from lanuching by throwing `GPUassert: no CUDA-capable device is detected`. Below is a minimal example on Perlmutter and my workaround.  It would be nice to not have to manually save this variable and reset it around DESC imports, especially when they are not at the top level.

```
>>> import os
>>> os.environ["CUDA_VISIBLE_DEVICES"]
'0,1,2,3'
>>> cuda_devices = os.environ.get("CUDA_VISIBLE_DEVICES")
>>> from desc.equilibrium import Equilibrium
>>> os.environ["CUDA_VISIBLE_DEVICES"]
''
>>> os.environ["CUDA_VISIBLE_DEVICES"] = cuda_devices
>>> os.environ["CUDA_VISIBLE_DEVICES"]
'0,1,2,3'
```

### PR DESCRIPTION
**Pull request by dpanici**
Resolves #2323

### CODE REVIEW

**Comment by dpanici** [comment]
```python
from desc import set_device
set_device("cpu")
from desc.backend import print_backend_info
print_backend_info()
```

this gives just the CPU when i print backend so I think it works fine? Also tested and torch seems to use the GPU as well after doing this on this PR

**Comment by YigitElma** [comment]
I tested this locally with GPU, it works. This also solves the issue that annoyed me for a while, on master,
```
>>> from desc import set_device
>>> set_device("cpu")
>>> import jax.numpy as jnp
>>> jnp.arange(8).device
ERROR:2026-09-15 00:04:27,046:jax._src.xla_bridge:487: Jax plugin configuration error: Exception when calling jax_plugins.xla_cuda12.initialize()
Traceback (most recent call last):
  File "/home/yigit/anaconda3/envs/gpu/lib/python3.13/site-packages/jax/_src/xla_bridge.py", line 485, in discover_pjrt_plugins
    plugin_module.initialize()
    ~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/yigit/anaconda3/envs/gpu/lib/python3.13/site-packages/jax_plugins/xla_cuda12/__init__.py", line 328, in initialize
    _check_cuda_versions(raise_on_first_error=True)
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yigit/anaconda3/envs/gpu/lib/python3.13/site-packages/jax_plugins/xla_cuda12/__init__.py", line 285, in _check_cuda_versions
    local_device_count = cuda_versions.cuda_device_count()
RuntimeError: jaxlib/cuda/versions_helpers.cc:113: operation cuInit(0) failed: CUDA_ERROR_NO_DEVICE
CpuDevice(id=0)
```
Here, I have `jax[cuda12]` installed, but using only CPU throws this loud error. With this change,
```
>>> from desc import set_device
>>> set_device("cpu")
>>> import jax.numpy as jnp
>>> jnp.arange(8).device
CpuDevice(id=0)
```

Couple things to note. `CUDA_VISIBLE_DEVICES` on a local system is not set by default, it is set on slurm clusters because same node has multiple GPUs, and slurm automatically blocks you from using them by setting this. So, I think most of our `set_device` function is redundant since Slurm already gives us 1 device to use (I assume you asked for 1 device). 

I guess in the case where a user requests multiple GPUs (so `CUDA_VISIBLE_DEVICES` is set and has more than 1 values), where only 1 of them is for DESC and the rest is for some other codes like the issue mentions, we should expect users to handle it without us making any changes. This will not change the behavior for cluster users who only request 1 GPU, but will probably make the life of users who want to use other codes easier.

TLDR; this change works, but I think we can go further and simplify the `set_device("gpu")` part too. The fancy-ish "assign most memory GPU" logic is redundant.

**Comment by YigitElma** [comment]
We should still test this on the cluster and see if JAX allocates the memory or not on the GPU.

**Comment by YigitElma** [comment]
@dpanici I made some changes to simplify the `set_device` process; you can revert them if you want to. These changes should allow:
- removing the `nvidia-ml-py` dependency
- using `tpu` backend
- letting users know if they call `set_device` in the wrong place

Only the users that have multiple GPUs visible to the Python process should be affected by this (previously we were picking 1 of them automatically), but I think it is safer for users to decide what they want to do with multiple devices

**Comment by dpanici** [comment]
Did you test local/on cluster? If so I am fine with the changes

**Comment by YigitElma** [comment]
> Did you test local/on cluster?

I tested locally for different scenarios. I can test it on the cluster later.

**Comment by dpanici** [comment]
I tested this and it seems to work on cluster too

File: desc/__init__.py [hunk 1]
```diff
@@ -65,127 +65,40 @@ def __getattr__(name):
...
-                    "yellow",
-                )
+    else:
+        os.environ.pop("JAX_PLATFORMS", None)
+        if kind == "gpu" and gpuid is not None:
+            # so that the ids assigned by CUDA match those from nvidia-smi
+            os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
+            visible = os.environ.get("CUDA_VISIBLE_DEVICES", "").split(",")
+            visible = [i for i in visible if i]
+            os.environ["CUDA_VISIBLE_DEVICES"] = (
```
Comment (YigitElma) [inline desc/__init__.py]: We can replace this with `JAX_CUDA_VISIBLE_DEVICES` such that only JAX-related visibility is affected but not a big deal since this is optional now.

Comment (dpanici) [inline desc/__init__.py]:
```suggestion
            os.environ["JAX_CUDA_VISIBLE_DEVICES"] = (
```

File: desc/__init__.py [hunk 2]
```diff
@@ -65,127 +65,40 @@ def __getattr__(name):
...
-                    "Specified gpuid {} not found, falling back to ".format(str(gpuid))
-                    + config["device"],
-                    "yellow",
-                )
+    else:
+        os.environ.pop("JAX_PLATFORMS", None)
+        if kind == "gpu" and gpuid is not None:
+            # so that the ids assigned by CUDA match those from nvidia-smi
+            os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
+            visible = os.environ.get("CUDA_VISIBLE_DEVICES", "").split(",")
```
Comment (dpanici) [inline desc/__init__.py]:
```suggestion
            visible = os.environ.get("JAX_CUDA_VISIBLE_DEVICES", "").split(",")
```

File: desc/__init__.py [hunk 3]
```diff
@@ -65,127 +65,40 @@ def __getattr__(name):
...
     Parameters
     ----------
-    kind : {``'cpu'``, ``'gpu'``}
-        whether to use CPU or GPU.
+    kind : {``'cpu'``, ``'gpu'``, ``'tpu'``}
+        Which device to use. On CPU, the accelerators are left untouched so that
+        other codes can use them.
+    gpuid : int, optional
+        Index of the GPU to use, in ``nvidia-smi`` ordering, or an index into
+        ``CUDA_VISIBLE_DEVICES`` if that is set. If ``None``, JAX will use every
```
Comment (dpanici) [inline desc/__init__.py]:
```suggestion
        ``JAX_CUDA_VISIBLE_DEVICES`` if that is set. If ``None``, JAX will use every
```

**Review by ddudt — DISMISSED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -32,6 +32,8 @@ Bug Fixes
 - Stops `ProximalProjection` from mutating `solve_options` during iterations.
 - Fixed bug that occured when passing in ``_surf_batch_size`` kwarg to ``Omnigenity`` and ``QuasisymmetryBoozer`` objectives
 - Fixes ``pitch_batch_size`` argument getting ignored in compute functions.
+- Stopped wiping the ``CUDA_VISIBLE_DEVICES`` environment variable when ``set_device("cpu")`` is used, which should now allow DESC to be used on CPU concurrently with e.g. pytorch or some other code using the GPU.
+
```
Comment (ddudt) [inline CHANGELOG.md]:
```suggestion
```

**Review by ddudt — APPROVED** [review]

**Comment by kvnloo** [comment]
I took a pass through this and noticed one edge case around `CUDA_VISIBLE_DEVICES="all"`.

JAX accepts `"all"` here, but the current selection path appears to treat the value like a comma-separated list of device IDs. That gives two different failure modes:

- requesting GPU 0 leaves visibility unrestricted rather than selecting GPU 0
- requesting GPU 1 can hit an `IndexError`

I tested a minimal normalization that treats `"all"` as the full available-device set before applying the requested `gpuid`. In an isolated test of this path I get 17/17 passing versus 15/17 before the change.

Would it make sense to handle `"all"` explicitly here? Happy to share the small patch/test if useful.

**Comment by dpanici** [comment]
@kvnloo Sure, please share the test/patch in a comment here, I am unfamiliar with this edge case

**Comment by kvnloo** [comment]
Thanks @dpanici, I rechecked this against the current head `ff2ef7f`.

Could I sanity-check how you want `JAX_CUDA_VISIBLE_DEVICES=all` to interact with an explicit `gpuid`?

Right now this path:

```python
visible = os.environ.get("JAX_CUDA_VISIBLE_DEVICES", "").split(",")
visible = [i for i in visible if i]
os.environ["JAX_CUDA_VISIBLE_DEVICES"] = (
    visible[int(gpuid)] if visible else str(int(gpuid))
)
```

turns `all` into a one-element list containing `all`. So from my focused checks:

- calling `set_device` with `kind="gpu"` and `gpuid=0` leaves `JAX_CUDA_VISIBLE_DEVICES=all`
- calling `set_device` with `kind="gpu"` and `gpuid=1` tries to access `visible[1]` and raises `IndexError`

Would the intended semantics be that `all` behaves like having no existing restriction when the user explicitly supplies a `gpuid`?

If so, the smallest change I found was:

```python
visible = os.environ.get("JAX_CUDA_VISIBLE_DEVICES", "").split(",")
visible = [i for i in visible if i and i != "all"]
```

which would make an explicit `gpuid` select the corresponding physical GPU in this case.

I was thinking the regression cases could then cover `all` with both `gpuid=0` and `gpuid=1`.

Does that interpretation match how you want `set_device` to behave, or is there another meaning for `all` that we should preserve here?

AI-assisted investigation; I only ran focused selection-path checks, not the full GPU suite.

**Comment by kvnloo** [comment]
hey, actually @dpanici could I check one additional boundary before treating the normalization as sufficient?

My earlier tests checked the environment write. In a fresh process with installed JAX `0.9.0.1`, the environment value doesn't appear to reach JAX's configuration:

```sh
JAX_CUDA_VISIBLE_DEVICES=1 python - <<'PY'
import os
import jax
print(jax.__version__, os.environ["JAX_CUDA_VISIBLE_DEVICES"])
print(jax.config.read("jax_cuda_visible_devices"))
jax.config.update("jax_cuda_visible_devices", "1")
print(jax.config.read("jax_cuda_visible_devices"))
PY
```

This prints `0.9.0.1 1`, then `all`, then `1`. A separate `JAX_ENABLE_X64` environment control works in the same fresh-process setup.

Am I missing a DESC initialization hook that applies the visibility setting to JAX? If not, would an explicit JAX configuration update before the first backend operation fit this PR's goal of leaving CUDA-wide visibility untouched?

The one-line normalization fixes the two source-level parsing cases, but this suggests those checks alone aren't enough to validate the handoff.

AI-assisted; executed fresh-process JAX configuration checks, not GPU selection or the full DESC suite.
